arch/x86/core/pcie.c

```diff
@@ -214,14 +219,15 @@ uint16_t pcie_msi_mdr(unsigned int irq,
 
 	if (vector == NULL) {
 		/* edge triggered */
-		return 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+		msg_data_reg = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+	} else if (!IS_ENABLED(CONFIG_INTEL_VTD_ICTL)) {
+		msg_data_reg = vector->arch.vector;
```

Comment:
This is a fix that seems to be related to VT-d, not MSI-X?   Again, can we get some clarity as to what's going on here?

Reply:
This added to keep previous logic for VT-d and legacy interrupt intact . Also for MSI/MSI-X the irq number is not relevant (please see my previous comment) and  hence for MSI we cannot use Z_IRQ_TO_INTERRUPT_VECTOR() macro. 
But I believe it is better remove the MSI-X check so that the API honor both MSI and MSI-X equally as show bellow:
uint16_t pcie_msi_mdr(unsigned int irq,
		      msi_vector_t *vector)
{
	uint16_t msg_data_reg = 0;

	if ((vector != NULL) && (IS_ENABLED(CONFIG_PCIE_MSI) || 
			IS_ENABLED(CONFIG_PCIE_MSI_X) ) {
		msg_data_reg  = vector->arch.vector;
	}
	else if (vector == NULL)) {
		/* edge triggered */
		msg_data_reg  = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
	}

	/* VT-D requires it to be 0, so let's return 0 by default */
	return msg_data_reg;
}